### PR TITLE
Match simGCN.c

### PR DESCRIPTION
**Pull request by Prakxo**

### CODE REVIEW

File: src/emulator/simGCN.c [hunk 1]
```diff
@@ -90,153 +77,15 @@ s16 Vert_s16Bar[12] ALIGNAS(32) = {
 };
 
 u32 Colors_u32[3] ALIGNAS(32) = {
-    0x000000FF,
-    0x000000FF,
-    0x000000FF,
+    255,
+    255,
+    255,
```
Comment (cadmic) [inline src/emulator/simGCN.c:82]: Hmm is this a single white color, or is it an array of 3 solid blacks?

File: src/emulator/simGCN.c [hunk 2]
```diff
@@ -641,7 +1734,124 @@ bool simulatorCopyControllerMap(u32* mapDataOutput, u32* mapDataInput) {
...
+        }
+        if ((subStickX != 0) && (subStickY != 0)) {
+            subStickTest = (f32)subStickX / (f32)subStickY;
+        } else {
+            subStickTest = 0.0f;
+        }
+        if (subStickTest < 0.0f) {
+            subStickTest *= -1.0f;
+        }
+        if ((subStickTest > 1.1f) || (subStickTest < 0.9090908765792847f)) {
```
Comment (cadmic) [inline src/emulator/simGCN.c:1801]: is this (1 / 1.1f)?

File: src/emulator/simGCN.c [hunk 3]
```diff
@@ -467,11 +280,112 @@ bool simulatorGXInit(void) {
...
+        if (pal->descriptorArray[i].CLUTHeader) {
+            pal->descriptorArray[i].CLUTHeader = (CLUTHeaderPtr)((u8*)pal + (u32)pal->descriptorArray[i].CLUTHeader);
+            if (!pal->descriptorArray[i].CLUTHeader->unpacked) {
+                pal->descriptorArray[i].CLUTHeader->data = (char*)pal + (u32)pal->descriptorArray[i].CLUTHeader->data;
+                pal->descriptorArray[i].CLUTHeader->unpacked = true;
+            }
+        }
+    }
+}
+bool simulatorDrawErrorMessage(__anon_0x61D7 simulatorErrorMessage, s32 drawBar, s32 percent);
```
Comment (cadmic) [inline src/emulator/simGCN.c:307]: hm I think this declaration looks a little awkward here, I'd move it either either to the .h file (since it's not static) or near the top (like they are in most of the .c files)

File: src/emulator/simGCN.c [hunk 4]
```diff
@@ -467,11 +280,112 @@ bool simulatorGXInit(void) {
...
+        if ((nStatus != 1) && (nStatus != 0) && (nStatus != 2) && (nStatus != 3) && (nStatus != 7) && (nStatus != 8) &&
+            (nStatus != 10)) {
+            toggle = true;
+        } else if (toggle == true) {
+            toggle = false;
+            nMessage = S_M_DISK_READING_DISK;
+        }
+
+        if ((gDVDResetToggle == 1) && ((nStatus <= 3U) || ((nStatus - 7) <= 1U) || (nStatus == 10))) {
+            if (simulatorTestReset(false, false, true, false) == false) {
```
Comment (cadmic) [inline src/emulator/simGCN.c:368]: `if (f() == false)` -> `if (!f())` (ditto 2x below)

File: src/emulator/simGCN.c [hunk 5]
```diff
@@ -467,11 +280,112 @@ bool simulatorGXInit(void) {
...
+                return false;
+                ;
```
Comment (cadmic) [inline src/emulator/simGCN.c:370]: extra semicolon (ditto below, and for the while loop `while (!frameBeginOK(gpSystem->pFrame)) {}` is better

File: src/emulator/simGCN.c [hunk 6]
```diff
@@ -519,25 +433,1204 @@ bool simulatorPlayMovie(void) {
...
+    f32 g2DviewMtx[3][4] = {
+        {1.0, 0.0, 0.0, 0.0},
+        {0.0, 1.0, 0.0, 0.0},
+        {0.0, 0.0, 1.0, -1.0},
+    };
+
+    f32 g2[3][4];
+
+    do {
+    } while (frameBeginOK(gpFrame) != 1);
```
Comment (cadmic) [inline src/emulator/simGCN.c:455]: `while (frameBeginOK(gpFrame) != true) {}` (ditto below)

File: src/emulator/simGCN.c [hunk 7]
```diff
@@ -378,7 +191,7 @@ bool simulatorGXInit(void) {
     GXLoadPosMtxImm(identity_mtx, 0);
     GXLoadNrmMtxImm(identity_mtx, 0);
     GXSetCurrentMtx(0);
-    GXLoadTexMtxImm(identity_mtx, 0x3C, 0);
+    GXLoadTexMtxImm(identity_mtx, 60, 0);
```
Comment (cadmic) [inline src/emulator/simGCN.c:194]: I'm not familiar with these functions, is this related to the `0x3C`s just above?

File: src/emulator/simGCN.c [hunk 8]
```diff
@@ -519,25 +433,1204 @@ bool simulatorPlayMovie(void) {
...
+    gbDisplayedError = true;
+    return true;
+}
+
+s32 simulatorPrepareMessage(__anon_0x61D7 simulatorErrorMessage) {
+    DVDFileInfo fileInfo;
+    switch (simulatorErrorMessage) {
+
+        case S_M_CARD_IN02:
+
```
Comment (cadmic) [inline src/emulator/simGCN.c:1080]: nit: the line breaks are a bit inconsistent in this function

File: src/emulator/simGCN.c [hunk 9]
```diff
@@ -519,25 +433,1204 @@ bool simulatorPlayMovie(void) {
...
+            gButtonDownToggle = true;
+            return true;
+        }
+        if (DemoPad->pst.button & 0x200) {
+            soundPlayBeep(SYSTEM_SOUND(gpSystem), SOUND_BEEP_DECLINE);
+            gButtonDownToggle = true;
+            *yes = 0;
+            return true;
+        }
+        if (((s8)(u8)DemoPad->pst.stickX < 0) && ((s32)*yes == 0)) {
```
Comment (cadmic) [inline src/emulator/simGCN.c:1179]: are all these casts necessary? I think `*yes` is already `s32` and `stickX` is already `s8`

File: src/emulator/simGCN.c [hunk 10]
```diff
@@ -519,25 +433,1204 @@ bool simulatorPlayMovie(void) {
...
+    }
+
+    return false;
+}
+
+bool simulatorDrawYesNoMessage(__anon_0x61D7 simulatorMessage, s32* yes) {
+    DVDFileInfo fileInfo;
+    switch (simulatorMessage) {
+        case S_M_CARD_LD05_2:
+
```
Comment (cadmic) [inline src/emulator/simGCN.c:1195]: nit: more line breaks inconsistency (sorry)

File: src/emulator/simGCN.c [hunk 11]
```diff
@@ -641,7 +1734,124 @@ bool simulatorCopyControllerMap(u32* mapDataOutput, u32* mapDataInput) {
...
+        *anData |= (stickX & 0xFF) << 8;
+        UnkInlineReadController(DemoPad[channel].pst.stickY, &stickY);
+        *anData |= stickY & 0xFF;
+
+        nDirButton = DemoPad[channel].dirs;
+        subStickX = DemoPad[channel].pst.substickX;
+        subStickY = DemoPad[channel].pst.substickY;
+        if (gButtonDownToggle == true) {
+            if (DemoPad->pst.button != 0) {
+                return 1;
```
Comment (cadmic) [inline src/emulator/simGCN.c:1782]: `return true;` (also I think the function probably returns `bool`)

File: src/emulator/simGCN.c [hunk 12]
```diff
@@ -997,9 +2247,15 @@ bool xlMain(void) {
     if (!xlHeapGetFree(&nSize0)) {
         return false;
     }
+
+    (void)0.0f;
+    (void)1.0f;
+
+    (void)0.1f;
```
Comment (cadmic) [inline src/emulator/simGCN.c:2254]: maybe add a comment here on why these are necessary?

File: src/emulator/simGCN.c [hunk 13]
```diff
@@ -519,25 +433,1204 @@ bool simulatorPlayMovie(void) {
...
+    return false;
+}
+
+static inline bool simulatorDrawOKMessageLoop(TEXPalettePtr simulatorMessage) {
+    simulatorDrawOKImage(
+        (TEXPalettePtr)gpErrorMessageBuffer,
+        N64_FRAME_WIDTH / 2 - ((TEXPalettePtr)gpErrorMessageBuffer)->descriptorArray->textureHeader->width / 2,
+        N64_FRAME_HEIGHT / 2 - ((TEXPalettePtr)gpErrorMessageBuffer)->descriptorArray->textureHeader->height / 2,
+        simulatorMessage, N64_FRAME_WIDTH / 2 - simulatorMessage->descriptorArray->textureHeader->width / 2,
+        180 - ((TEXPalettePtr)gyes)->descriptorArray->textureHeader->height / 2);
```
Comment (cadmic) [inline src/emulator/simGCN.c:1299]: maybe comment on `gyes` here?

**Review by cadmic — COMMENTED** [review]
Awesome work, apologies for the nitpicks

File: src/emulator/simGCN.c [hunk 14]
```diff
@@ -519,25 +433,1204 @@ bool simulatorPlayMovie(void) {
...
+    f32 identity_mtx[3][4] = {
+        {1.0, 0.0, 0.0, 0.0},
+        {0.0, 1.0, 0.0, 0.0},
+        {0.0, 0.0, 1.0, -1.0},
+    };
+    f32 g2DviewMtx[3][4] = {
+        {1.0, 0.0, 0.0, 0.0},
+        {0.0, 1.0, 0.0, 0.0},
+        {0.0, 0.0, 1.0, -1.0},
+    };
```
Comment (Yanis42) [inline src/emulator/simGCN.c:450]: small detail but I'd add ``f`` to these floats (``1.0f, 0.0f`` etc), it seems there's other arrays like this below

File: src/emulator/simGCN.c [hunk 15]
```diff
@@ -519,25 +433,1204 @@ bool simulatorPlayMovie(void) {
...
-#pragma GLOBAL_ASM("asm/non_matchings/simGCN/simulatorDrawErrorMessage.s")
+bool simulatorDrawErrorMessageWait(__anon_0x61D7 simulatorErrorMessage) {
+    DVDFileInfo fileInfo;
 
-#pragma GLOBAL_ASM("asm/non_matchings/simGCN/simulatorPrepareMessage.s")
+    switch (simulatorErrorMessage) {
+        case S_M_CARD_LD01:
+            if (simulatorMessageCurrent != simulatorErrorMessage) {
+                simulatorMessageCurrent = simulatorErrorMessage;
+                if (DVDOpen("TPL/msg_ld01.tpl", &fileInfo) == 1) {
```
Comment (Yanis42) [inline src/emulator/simGCN.c:1335]: something I thought of earlier, should we add comments that says what's the message? for instance for this one it would be like ``// "A Memory Card is not inserted in Slot A. Please check to make sure that a Memory Card is inserted."``

File: src/emulator/simGCN.c [hunk 16]
```diff
@@ -641,7 +1734,124 @@ bool simulatorCopyControllerMap(u32* mapDataOutput, u32* mapDataInput) {
...
+    if (((stickValue >= 0) && (stickValue < 0x28)) || ((stickValue < 0) && (stickValue > -0x28))) {
+        ret = (stickValue * 0x43) / 40;
+    } else if ((stickValue >= 0x28) && (stickValue < 0x48)) {
+        ret = (((0x48 - stickValue) * 0x43) / 32) + (((stickValue - 0x28) * 0x5A) / 32);
+    } else if ((stickValue <= -0x28) && (stickValue > -0x48)) {
+        ret = (((-0x28 - stickValue) * -0x5A) / 32) + (((stickValue + 0x48) * -0x43) / 32);
+    } else if (stickValue >= 0x48) {
+        ret = 0x5A;
+    } else {
+        ret = -0x5AU;
```
Comment (Yanis42) [inline src/emulator/simGCN.c:1748]: is the U really necessary here?

File: src/emulator/simGCN.c (hunk 14 above)
Comment (Yanis42) [reply src/emulator/simGCN.c:450]: btw you might be able to use ``Mtx``, like ``Mtx identity_mtx = { {1.0f, 0.0f, 0.0f, 0.0f}, ... };``

File: src/emulator/simGCN.c (hunk 15 above)
Comment (Prakxo) [reply src/emulator/simGCN.c:1335]: Might be out of scope for this PR, since I can't really check lol

File: src/emulator/simGCN.c (hunk 7 above)
Comment (Prakxo) [reply src/emulator/simGCN.c:194]: these come from here https://github.com/doldecomp/dolsdk2001/blob/1f0c32f172557f868930bc68a31ac169ea6d9c69/src/gx/GXInit.c#L239C48-L239C57
I need to fix these

File: src/emulator/simGCN.c (hunk 1 above)
Comment (Prakxo) [reply src/emulator/simGCN.c:82]: I can't really tell
Not familiar with the functions that this gets passed to, but since they're supposedly colour values, I put the values in dec

File: src/emulator/simGCN.c (hunk 15 above)
Comment (Yanis42) [reply src/emulator/simGCN.c:1335]: it's fine if it's not done now, I just thought of that and wanted to talk about it for later

File: src/emulator/simGCN.c [hunk 17]
```diff
@@ -1296,7 +1288,7 @@ static inline bool simulatorDrawOKMessageLoop(TEXPalettePtr simulatorMessage) {
         N64_FRAME_WIDTH / 2 - ((TEXPalettePtr)gpErrorMessageBuffer)->descriptorArray->textureHeader->width / 2,
         N64_FRAME_HEIGHT / 2 - ((TEXPalettePtr)gpErrorMessageBuffer)->descriptorArray->textureHeader->height / 2,
         simulatorMessage, N64_FRAME_WIDTH / 2 - simulatorMessage->descriptorArray->textureHeader->width / 2,
-        180 - ((TEXPalettePtr)gyes)->descriptorArray->textureHeader->height / 2);
+        180 - ((TEXPalettePtr)gyes)->descriptorArray->textureHeader->height / 2); //bug, copy paste error?
```
Comment (cadmic) [inline src/emulator/simGCN.c:1291]: super nit: add space between `//` and `bug` (also, reformat file?)

File: src/emulator/simGCN.c (hunk 1 above)
Comment (cadmic) [reply src/emulator/simGCN.c:82]: hm just guessing based off these lines:
```
GXSetArray(GX_VA_CLR0, Colors_u32, 4);
GXSetVtxAttrFmt(GX_VTXFMT0, GX_VA_CLR0, GX_CLR_RGBA, GX_RGB8, 0);
```
I think these are 32-bit RGBA values as fully-opaque black like `rgba(0, 0, 0, 255)`. So I think 0x000000FF is clearer, but it's not a big deal

(I personally prefer hex for colors always, but that's not consistent right now)

File: src/emulator/simGCN.c (hunk 7 above)
Comment (cadmic) [reply src/emulator/simGCN.c]: thanks for cleaning that up!

**Review by Yanis42 — APPROVED** [review]

**Review by cadmic — APPROVED** [review]
